### ISSUE TITLE
RealFP types too restrictive for signalcorr.jl methods

### ISSUE THREAD

**Issue by danmackinlay** [JuliaStats/StatsBase.jl#459]
the methods in `signcorr.jl` require arrays of `RealFP` types;
This seems to me to be too restrictive since it, for example, the Dual types used by ForwardDiff.jl, and prevents differentiating through these functions.

Does anyone know a reason that we have this restriction? If I can get some clues about the reasons for this design choice I can propose a patch.

```julia
julia> using StatsBase:  autocov
julia> using ForwardDiff
julia> f2(x::Vector) = autocov(x, [0])[1]
julia> g2 = x -> ForwardDiff.gradient(f2, x);
julia> f2([1.0]), g2([1.0])

MethodError: no method matching fptype(::Type{ForwardDiff.Dual{ForwardDiff.Tag{typeof(f2),Float64},Float64,1}})
Closest candidates are:
  fptype(!Matched::Type{Complex{Float32}}) at /home/dan/.julia/packages/StatsBase/yNoh6/src/common.jl:27
  fptype(!Matched::Type{Complex{Float64}}) at /home/dan/.julia/packages/StatsBase/yNoh6/src/common.jl:28
  fptype(!Matched::Type{T<:Union{Bool, Float32, Int16, Int8, UInt16, UInt8}}) where T<:Union{Bool, Float32, Int16, Int8, UInt16, UInt8} at /home/dan/.julia/packages/StatsBase/yNoh6/src/common.jl:25
  ...
```

**Comment by nalimilan** [comment]
This code is quite old, it probably just hasn't been updated. As you can see, the error is due to the lack of an appropriate `fptype` method). I guess we could add a fallback definition `fptype(::Type{T}) where {T} = float(T)`. Does that fix your problem? If so, feel free to make a pull request.

**Comment by danmackinlay** [comment]
Hm - possibly. Thanks @nalimilan. I will check that for ForwardDiff. It will for sure help with things like BigFloats, which is another limitation the current code has.

Just so I can understand here, though, what we are doing with these `fptype` methods is trying to find the smallest float type such that there will be no precision lost when the covariance calculation is conducted in float arithmetic. Is that right?

If so, that looks like the [promotion methods built in to julia](https://docs.julialang.org/en/v1.1/manual/conversion-and-promotion/#Promotion-1). Is there a reason *not* to use those built in promotion methods? `signalcorr.jl` doesn't do anything very exotic, so the custom promotion logic is a bit weird.

**Comment by nalimilan** [comment]
Yes, actually `fptype` allows for more precision loss than `promote_type`. But here there's no other type with which to promote, so we should just use `float(eltype(X))`. It generally returns `Float64` instead of a smaller type, but at least that's more standard than using custom rules just for one function.

Anyway, even if you use a small input type, you may want to perform computations in `Float64` to avoid precision loss. And the standard way to choose a smaller type is to allocate the result manually and call `autocov!` on it.

**Comment by nalimilan** [comment]
Fixed by #567.
